### PR TITLE
[Feature] Add Ajax support to Quick button 😎

### PR DESCRIPTION
**Pull request by karandatwani92**
## WHY

The quick button is good for us(lazy devs), but Ajax support was missing.

Creating an Ajax button manually takes time. What if one project needs multiple Ajax buttons?
Everyone loves seamless UI(but with less effort).

### AFTER 
Now, We can create buttons quickly that support Ajax calling. Just add one attribute to have it.
- On success: it shows a notification.
- On Failure: it shows an alert.

https://github.com/Laravel-Backpack/CRUD/assets/8214221/54ac49b9-1b78-45fe-a3b1-d69d375a27f5

### Is it a breaking change?

NO

### How can we test the before & after?
It picks the URL from the same old href attribute, which was either user-provided or autogenerated.

Example:
I created an operation and used the following button to send invoices.
```php
CRUD::button('send')->stack('line')->view('crud::buttons.quick')->meta([
            'access' => true,
            'label' => 'Send Invoice',
            'icon' => 'la la-envelope',
            'ajax' => true,
        ]);
```

**OR**

```php
CRUD::button('send')->stack('line')->view('crud::buttons.quick')->meta([
            'access' => true,
            'label' => 'Send Invoice',
            'icon' => 'la la-envelope',
            'wrapper' => [
                'href' => url(('admin/dashboard')), // to change URL(nothing new)
            ],
            'ajax' => [
               // optional attributes
                'method' => 'GET',
                'refreshCrudTable' => true, // to refresh table on success
                'success_title' => "asas\'sds\"",
                'success_message' => 'The payment reminder has been sent successfully.',
                'error_title' => 'Error',
                'error_message' => 'There was an error sending the payment reminder. Please try again.',
            ],
        ]);
```

### CODE REVIEW

**Comment by tabacitu** [comment]
@pxpm will you please give this a review / polish? I think it's a small but cool thing we can launch by the end of the month.

**Comment by karandatwani92** [comment]
Hey @tabacitu @pxpm 

I have my opinion here. Why I chose:
- sweetalert on error.
- bubble notification on success.

Imagine I'm the admin and want to send payment reminders to users with the due amount. I filtered the users on CRUD.
Now I simply want to click click click on all rows. If everything is going smoothly, success bubble notifications rain without interrupting me.

If some row click in the middle has a problem causing an error, the sweet alert will interrupt and stop my finger click reflex action on the next row & I'll be able to remember what I clicked last, identifying which row had the problem.

But if we implement bubble notification on error, it will not stop my finger click reflex action and the error will be mixed in the raining notification, and it would be hard for me to identify the row with the error. Bubble notification also dismisses itself, so I might now be able to read the info in the short visible timespan. 

But with the error on sweet alert, I can patiently read the error and proceed further.

File: src/resources/views/crud/buttons/quick.blade.php [hunk 1]
```diff
@@ -37,3 +55,75 @@
...
+
+@if($ajax_enabled)
+{{-- Button Javascript --}}
+{{-- - used right away in AJAX operations (ex: List) --}}
+{{-- - pushed to the end of the page, after jQuery is loaded, for non-AJAX operations (ex: Show) --}}
+@push('after_scripts') @if (request()->ajax()) @endpush @endif
+@bassetBlock('backpack/crud/buttons/quick-button-'.app()->getLocale().'.js')
+<script>
+	if (typeof sendQuickRequest != 'function') {
+	  $("[data-button-type=quick]").unbind('click');
```
Comment (pxpm) [inline src/resources/views/crud/buttons/quick.blade.php:68]: This will unbind the clicks of all `quick` buttons, even the ones that are not ajax, won't that be a problem since the non-ajax buttons have no fallback to this function?

**Review by pxpm — CHANGES_REQUESTED** [review]
Hey @karandatwani92, good job 🙏 

I haven't tested this yet in actual scenarios, lets first clarify some points and then we move on. 

I've left some inline comments, and I still have some questions/concerns. 

I agree with Cristian in regard to alert for sucess/swal for error. I think you are trying to push a UX that's not very good. The example you gave is a perfect example of when you should use bulk actions, not click on every row. Just filter what you want, select all and "Do something" only once. 

There are also two questions that I think they will come up when we use this in a variety of scenarios, eg:

**1)** How can I reload the datatable after the ajax response ? Should I be able to ? I think yes right ? It's what happens for example in the Trash button. https://github.com/Laravel-Backpack/PRO/blob/edf0f8670eeb8f9dbd102937bcff380af197e747/resources/views/buttons/trash.blade.php#L47

**2)** Can I have custom messages for my users or can I just return `success/error` status code from the endpoint? 

Let me know what you think. 

Cheers

File: src/resources/views/crud/buttons/quick.blade.php [hunk 2]
```diff
@@ -37,3 +55,75 @@
...
+		              	buttons: false,
+		              });
+			      }
+			  });
+
+      }
+	}
+
+	// make it so that the function above is run after each DataTable draw event
+	// crud.addFunctionToDataTablesDrawEventQueue('sendQuickRequest');
```
[the remark this answers is not in the document]
Comment (karandatwani92) [reply src/resources/views/crud/buttons/quick.blade.php:125]: Removing comment!

File: src/resources/views/crud/buttons/quick.blade.php (hunk 1 above)
Comment (karandatwani92) [reply src/resources/views/crud/buttons/quick.blade.php:68]: OK, but why is it present in the [delete button](https://github.com/Laravel-Backpack/CRUD/blob/f2257fe6e0da458bd9d170d353b15ec66a2ad254/src/resources/views/crud/buttons/delete.blade.php#L15)? 

It was there, so I added it.

File: src/resources/views/crud/buttons/quick.blade.php [hunk 3]
```diff
@@ -37,3 +55,75 @@
...
     </{{ $wrapper['element'] }}>
 @endif
+
+
+@if($ajax_enabled)
+{{-- Button Javascript --}}
+{{-- - used right away in AJAX operations (ex: List) --}}
+{{-- - pushed to the end of the page, after jQuery is loaded, for non-AJAX operations (ex: Show) --}}
+@push('after_scripts') @if (request()->ajax()) @endpush @endif
+@bassetBlock('backpack/crud/buttons/quick-button-'.app()->getLocale().'.js')
```
[the remark this answers is not in the document]
Comment (karandatwani92) [reply src/resources/views/crud/buttons/quick.blade.php:65]: Changing to `@bassetBlock('backpack/crud/buttons/quick-button.js')`

File: src/resources/views/crud/buttons/quick.blade.php [hunk 4]
```diff
@@ -23,6 +23,24 @@
         $wrapper['href'] = ($wrapper['href'])($entry, $crud);
     }
     $wrapper['class'] = $wrapper['class'] ?? $defaultClass;
+    //if ajax enabled
+    $ajax_enabled = $button->meta['ajax'] ?? false;
```
[the remark this answers is not in the document]
Comment (karandatwani92) [reply src/resources/views/crud/buttons/quick.blade.php:27]: My Initial thought:
If so, why are we not doing the same for existing operation buttons?
Let's say if we do for all - But no. of file will increase.

**Then my brain argues**, They are not two-in-one buttons like this. If they are the Ajax buttons, they are only Ajax buttons, so the **files will not increase**.

My final thought is that I like to keep this tiny thing in a single file; I don't like jumping on files for this small thing.

So it's a preference thing. Let me know your final command; I will walk in your direction, Master :)

**Comment by karandatwani92** [comment]
Hey @pxpm 

Thanks for the feedback; I have left a few comments and resolved a few as directed. Now, coming to major questions:
1) How can I reload the data table after the Ajax response?
-- Nice idea, but I wouldn't say I like the whole data table reload. A single-element reload would look nice. What do you think about it & how would you implement it?

2) Can I have custom messages for my users or return the success/error status code from the endpoint?
-- I just pushed a few commits for this. If `abort()` or some exception, it will display an error message from the server. If the success message is passed, it will display the server response message; if unavailable - it will display the default message. If one uses some external API and can't change the response, he can set success/error messages via the field attribute.

File: src/resources/views/crud/buttons/quick.blade.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (pxpm) [reply src/resources/views/crud/buttons/quick.blade.php:27]: I usually agree with that, but in this case, and like you pointed out, this represent two buttons in one. 
It's not that we have plans to re-use the ajax part (theoretically we could) but this 40 line file, went to 100 lines, so more than half of it is now the ajax functionality.

In this case I still think it's better to move the ajax part to a separate file, it will help to keep both files clean and easy to understand. 

We can check later with Cristian about his opinion on this, it's not something that's preventing us from moving forward, if we decide to do it, there shouldn't be any code changes needed, so we leave this to decide at the end 👍

File: src/resources/views/crud/buttons/quick.blade.php [hunk 5]
```diff
@@ -37,3 +55,44 @@
...
+{{-- Button Javascript --}}
+{{-- Pushed to the end of the page, after jQuery is loaded --}}
+@push('after_scripts') @if (request()->ajax()) @endpush @endif
+@bassetBlock('backpack/crud/buttons/quick-button.js')
+<script>
+	if (typeof sendQuickButtonAjaxRequest != 'function') {
+	  $("[data-button-type=quick]").unbind('click');
+
+	  function sendQuickButtonAjaxRequest(button) {
+		// e.preventDefault();
```
Comment (pxpm) [inline src/resources/views/crud/buttons/quick.blade.php:70]:
```suggestion
```

File: src/resources/views/crud/buttons/quick.blade.php (hunk 4 above)
Comment (pxpm) [inline src/resources/views/crud/buttons/quick.blade.php:27]: Maybe we add an additional check here to avoid syntax errors ? 
```suggestion
    $buttonAjaxConfiguration= $button->meta['ajax'] ?? false;
    if($buttonAjaxConfiguration && !is_array($buttonAjaxConfiguration))   {
        $buttonAjaxConfiguration = false;
     }
```

Comment (pxpm) [reply src/resources/views/crud/buttons/quick.blade.php:27]: I've also renamed the variable, I think it's more clear what it holds. 👍

**Comment by pxpm** [comment]
> How can I reload the data table after the Ajax response?
-- Nice idea, but I wouldn't say I like the whole data table reload. A single-element reload would look nice. What do you think about it & how would you implement it?

I think single element is not something achievable in datatables. What we do as I've showed in trash, is that we check if there is only one item in the current table, if we have only one we will move to the previous page, and after that we call the datatable redraw. But that only works in trash and delete, that we are sure the entry is deleted/removed from the dataset. 

In this case, most likely the entry will not be deleted, so I think we shouldn't move to the previous page and only call the redraw. If needed in the future we can add that option like `removesEntryFromTable => true/false`

I think an "easy-way" to add the reload option without much bell and whistles is to add an option like:
`refreshCrudTable` - true/false, in case true we run our reload table script, it can be false by default.

**Comment by pxpm** [comment]
Another thing that may come handy and may be requested, is a way to "wait for request completion", something like a loading state, instead of allowing the user to send multiple requests at once. 

This can be worked out later if needed, I just wanted to let this written here. 👍

**Comment by karandatwani92** [comment]
> removesEntryFromTable

Optional attribute added:
`removesEntryFromTable => true/false`

File: src/resources/views/crud/buttons/quick.blade.php (hunk 5 above)
Comment (karandatwani92) [reply src/resources/views/crud/buttons/quick.blade.php]: Comment removed

File: src/resources/views/crud/buttons/quick.blade.php [hunk 6]
```diff
@@ -37,3 +55,44 @@
...
+		var route = $(button).attr('data-route');
+				$.ajax({
+			      url: route,
+			      type: $(button).attr('data-method'),
+			      success: function(result) {
+			        new Noty({
+		            	type: "success",
+						text: result.message?result.message:'<strong>'+$(button).attr('data-success-title')+'</strong><br>'+$(button).attr('data-success-message'),
```
[the remark this answers is not in the document]
Comment (karandatwani92) [reply src/resources/views/crud/buttons/quick.blade.php:81]: Again, it's a preference. One-liner is good. It's not a complicated-looking ternary operation.

I see the same practice in other buttons.

File: src/resources/views/crud/buttons/quick.blade.php (hunk 1 above)
Comment (karandatwani92) [reply src/resources/views/crud/buttons/quick.blade.php:68]: Ok... But this `$wrapper['data-button-type'] = 'quick'`  is only attached when the button has the `ajax` attribute set. So it will not affect other non-ajax quick buttons.

Tell me your final words, I'll do it... I am just trying to understand why `unbind` is there on the delete button which made me copy it here. 😅

This discussion is just to learn and not repeat later. Thanks, Master

Comment (pxpm) [reply src/resources/views/crud/buttons/quick.blade.php:68]: Disregard then, in that case it's not affecting the other buttons like you said. 

I thought that was an attribute on ALL quick buttons, since the name is a generic "quick", if it would be `['data-button-type'] = 'quick-ajax'` I wouldn't have this doubt. 

Maybe worth to rename it, we don't have the "need" for adding `button-type` to non ajax **now**, but in the future we may need. And if that happens we will need to come up with a different key or making a breaking change. 

Let's start then with `= 'quick-ajax'` as the button type.

File: src/resources/views/crud/buttons/quick.blade.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (pxpm) [reply src/resources/views/crud/buttons/quick.blade.php:81]: It still one liner, but at the moment there is so much to digest in that line that is very difficult to read.

- it's already a ternary operation, so not the most "expressive" one. 
- they keys are nested, I mean `result.message` instead of only `message`. 
- adding html to the string
- adding $() jquery wrappers and selectors

>I see the same practice in other buttons.

We are always learning. Don't let the errors from the past prevent you from doing better in future. 

I would lean to go even further to make it easier to read (those + plus signs still pollute the vision), using template literals:
```javascript
`<strong>${buttonSuccessTitle}</strong><br/>${buttonSuccessMessage}`;
```

I don't think in this case it's a "preference".
` result.message?result.message:'<strong>'+$(button).attr('data-success-title')+'</strong><br>'+$(button).attr('data-success-message'),` is not an easy to read string anywhere in the world.

File: src/resources/views/crud/buttons/quick.blade.php [hunk 7]
```diff
@@ -23,6 +23,25 @@
         $wrapper['href'] = ($wrapper['href'])($entry, $crud);
     }
     $wrapper['class'] = $wrapper['class'] ?? $defaultClass;
+    //if ajax enabled
+    $ajax_enabled = $button->meta['ajax'] ?? false;
+    if($ajax_enabled) {
+        $wrapper['data-route'] = $wrapper['href'];
+		$wrapper['data-method'] = $button->meta['ajax']['method'] ?? 'POST';
+        $wrapper['data-refresh-table'] = $button->meta['ajax']['removesEntryFromTable'] ?? false;
```
Comment (pxpm) [inline src/resources/views/crud/buttons/quick.blade.php:31]: `removeEntryFromTable` is not what this does. 
Sorry if I didn't make myself clear when I explained my expectations for a table refresh. 

I think there are 2 scenarios:
1 - you do something with the entry but don't remove it (Eg, changing the status)
2 - you do remove the entry (mark it as deleted, or actually delete it). 

I think for starters we should cater only to **1**. `data-refresh-table => refreshCrudTable ?? false`. 

We can introduce separately another configuration (**but I think we shouldn't do it now, only if we need or people ask for it**), like `removesEntryFromTable`. If developer enables this, we need to do the `is there only one item in the table ?` verification that we do on delete and trash. 
But let's not work on that now.

```suggestion
        $wrapper['data-refresh-table'] = $button->meta['ajax']['refreshCrudTable'] ?? false;
```

File: src/resources/views/crud/buttons/quick.blade.php (hunk 4 above)
Comment (pxpm) [reply src/resources/views/crud/buttons/quick.blade.php:27]: Hum .. sure you are right. It does not need to be an array, disregard that part. 

 Let's just rename the variable, either what I proposed, or `quickAjaxConfiguration` or something similar.

File: src/resources/views/crud/buttons/quick.blade.php (hunk 1 above)
Comment (karandatwani92) [reply src/resources/views/crud/buttons/quick.blade.php:68]: Renaming `quick` to `quick-ajax`.

File: src/resources/views/crud/buttons/quick.blade.php (hunk 4 above)
Comment (karandatwani92) [reply src/resources/views/crud/buttons/quick.blade.php:27]: Done!

File: src/resources/views/crud/buttons/quick.blade.php (hunk 7 above)
Comment (karandatwani92) [reply src/resources/views/crud/buttons/quick.blade.php:31]: I was expecting this. I intentionally named it `data-refresh-table`. 😬.

Done!

File: src/resources/views/crud/buttons/quick.blade.php [hunk 8]
```diff
@@ -37,3 +56,56 @@
...
+			        new Noty({
+		            	type: "success",
+						text: message,
+		            }).show();
+			      },
+			      error: function(result) {
+			          // Show an alert with the result
+			          swal({
+						title: $(button).attr('data-error-title'),
+	                    text: result.responseJSON.message?result.responseJSON.message:$(button).attr('data-error-message'),
```
Comment (pxpm) [inline src/resources/views/crud/buttons/quick.blade.php:98]:
```suggestion
                      let buttonErrorMessage = button.getAttribute('data-error-message');
                      buttonErrorMessage = result.responseJSON.message ? result.responseJSON.message : buttonErrorMessage;
			          // Show an alert with the result
			          swal({
						title: button.getAttribute('data-error-title'),
	                    text: buttonErrorMessage,
```

File: src/resources/views/crud/buttons/quick.blade.php [hunk 9]
```diff
@@ -37,3 +56,56 @@
...
+                    let message;
+                    if(result.message){
+                        //if message is returned from the API
+                        message = result.message;
+                    }else{
+                        //if message is not returned from the API
+                        let buttonSuccessTitle = button.getAttribute('data-success-title');
+                        let buttonSuccessMessage =  button.getAttribute('data-success-message');
+                        message = `<strong>${buttonSuccessTitle}</strong><br/>${buttonSuccessMessage}`;
+                    }
```
Comment (pxpm) [inline src/resources/views/crud/buttons/quick.blade.php:88]:
```suggestion
                    let buttonSuccessTitle = button.getAttribute('data-success-title');
                    let buttonSuccessMessage =  button.getAttribute('data-success-message');
                    let message = `<strong>${buttonSuccessTitle}</strong><br/>${buttonSuccessMessage}`
                    if(result.message){
                        // if message is returned from the API use it instead of the default message
                        message = result.message;
                    }
                    
```

**Review by pxpm — CHANGES_REQUESTED** [review]
Just suggested two small changes and I think in terms of semantics/features we are done here. 
Would you mind writing the docs for it ? 

Cheers

File: src/resources/views/crud/buttons/quick.blade.php (hunk 9 above)
Comment (karandatwani92) [reply src/resources/views/crud/buttons/quick.blade.php:88]: @pxpm I intentionally went that way to save browser processing, to make it lighter.

If the result came from API, the Browser will not have to execute:
```javascript
let buttonSuccessTitle = button.getAttribute('data-success-title');
let buttonSuccessMessage =  button.getAttribute('data-success-message');
let message = `<strong>${buttonSuccessTitle}</strong><br/>${buttonSuccessMessage}`
```

Comment (pxpm) [reply src/resources/views/crud/buttons/quick.blade.php:88]: Sure, there is nothing wrong with your approach. 
I wouldn't have worried about saving this work from the browser. The "hard-work" is already done, here we have an html element (button object) and we just get some properties of that object to write a string. It's very very cheap work

If we were sending a request to the server or something similar, I would avoid making that request unless strictly necessary but in this case ... méh. 

But since we are exploring possibilities, how about this then:
```suggestion
                    let message;
                    let defaultMessage = function(button) {
                        // since this is inside a closure, we don't execute until we manually call it.
                        let buttonSuccessTitle = button.getAttribute('data-success-title');
                        let buttonSuccessMessage =  button.getAttribute('data-success-message');
                        return `<strong>${buttonSuccessTitle}</strong><br/>${buttonSuccessMessage}`;
                    }
                    
                    //if message is returned from the API use that message
                    if(result.message){
                        message = result.message;
                    }
                    
                    // if message variable has no value, we build the default message, otherwise use the previous set value without calling defaultMessage()
                    message ??= defaultMessage(button);
```

**Comment by karandatwani92** [comment]
**Docs PR:** https://github.com/Laravel-Backpack/docs/pull/581

File: src/resources/views/crud/buttons/quick.blade.php (hunk 9 above)
Comment (karandatwani92) [reply src/resources/views/crud/buttons/quick.blade.php:88]: Here, our objective is to make the code look simpler.

Let's see it with the eyes of a beginner. Do you think a closure looks simple? I feel that concatenation looks simpler.

Comment (pxpm) [reply src/resources/views/crud/buttons/quick.blade.php:88]: I don't get sorry, it's because I called it `closure` and not `function` ? Basically it's what it is, a **function** (a block of code you create and defer to execute at any given time in the future), in this case, an **anonymous function**, where you don't need to define visibility or names. Just store the function in a variable, and invoke it if needed. 

This would be pretty much the same as having: 
```javascript
function defaultMessage(button) {
// ...
}

message ??=defaultMessage(button); 
```

The usage of a closure allow us to don't polute the global space with "specific functions" that only belong to this bit of javascript. 

I am sorry if this doesn't feel good for a beginner, but this is as beginner as it can get.

File: src/resources/lang/en/crud.php [hunk 10]
```diff
@@ -190,4 +190,10 @@
 
     // The pivot selector required validation message
     'pivot_selector_required_validation_message' => 'The pivot field is required.',
+
+    // Quick button messages
+    'quick_button_ajax_error_title' => 'Request failed!',
+    'quick_button_ajax_error_message' => 'There was an error while executing request. Please retry.',
+    'quick_button_ajax_success_title' => 'Request successful!',
+    'quick_button_ajax_success_message' => 'The action request is submited successfully.',
```
[the remark this answers is not in the document]
Comment (karandatwani92) [reply src/resources/lang/en/crud.php:198]: They are going to be an async request.
- We are submitting an execution request `The action request is submitted successfully.`✅
- We don't get a final response, so it's not completed `Your request was completed with success.`❌

[the remark this answers is not in the document]
Comment (pxpm) [reply src/resources/lang/en/crud.php:198]: What's the case when you don't get a final response ? You always get the response for the request you initiate. 
You send a request to the server, server processes it and return the response. 

If you get a success status it means that:
- Your request was submitted to the server.
- Server processed you request. 
- Server returned a success response.

If you get an error status it means that:
- Your request was submitted to the server.
- Server processed you request. 
- Server returned an error response.

What I mean is that "submitted successfully" is "what happens always", be it success response or an error. You request is submitted with success on both cases, what differ is the response. 

So the intention here is to tell the user that whatever happened in that endpoint (send email, change status, ban the user ...), was completed with success. 

Calling it "action request" is personifying the request as an action, that in my opinion is already the point of the request. You NEVER send a request expecting nothing to happen, be it in browser or in api. 

Sorry if I missed it, can you rephrase in that case what you mean?

**Comment by pxpm** [comment]
> @pxpm will you please give this a review / polish? I think it's a small, quick, cool thing we can launch by the end of the month.
> 
> One note I see right away is that the notification text&type should probably NOT be defined in the blade file. But in the PHP that returns the response. That way, you could return different notifications depending on success/error. You might see other stuff, personally I only saw that one as pretty important and worth changing.
> 
> Wait no. I also saw that you have different notification types on success/error. On success you have bubble notification, on error sweetalert. No bueno - let's make them both regular bubble alerts - it's much simpler and good enough.
> 
> Anything else?

@karandatwani92 I think this is the only thing (apart from my comments) that needs to be addressed. 

I've also already comment on it on a previous iteration: 
> I agree with Cristian in regard to alert for success/swal for error. I think you are trying to push a UX that's not very good. The example you gave is a perfect example of when you should use bulk actions, not click on every row. Just filter what you want, select all and "Do something" only once.

I've read your thoughts on why you disagree with me and Cristian, but if that example is the only argument I don't think that will be enough to convince us to change our mind. 

If you can find a better use case for using both, let me know so that we can evaluate. 
As a counter to your argument, I would probably return a link to the entry on the notification, so that If I saw one red, I could just click the link on the notification and go to that entry for example. 

Let's finish this. 💪

**Comment by pxpm** [comment]
It's done! 
Thank you @karandatwani92 for all the effort here, I am really happy with how it ended and how it works. 🙏 

I just did a little cleanup here https://github.com/Laravel-Backpack/CRUD/pull/5512/commits/e1db6e3c137ac73527f54c51f1e473fa89efea7c

I think this completely drops my concern about separating the file into multiple files as this is now short and easy to reason about.

Good job 🥳
